src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,598 @@
+    storage_class: Option<S3StorageClass>,
+    tags: Option<BTreeMap<String, String>>,
+}
+
+impl GenerateConfig for DatadogArchivesSinkConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            service: "".to_owned(),
+            bucket: "".to_owned(),
+            key_prefix: None,
+            request: TowerRequestConfig::default(),
+            aws_s3: Some(S3Config::default()),
+        })
+        .unwrap()
+    }
+}
+
+#[derive(Debug, Snafu, PartialEq)]
+enum ConfigError {
+    #[snafu(display("Unsupported service: {}", service))]
+    UnsupportedService { service: String },
+    #[snafu(display("Unsupported storage class: {}", storage_class))]
+    UnsupportedStorageClass { storage_class: String },
+}
+
+const KEY_TEMPLATE: &str = "/dt=%Y%m%d/hour=%H/";
+
+impl DatadogArchivesSinkConfig {
+    fn new(&self, cx: SinkContext) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        match &self.service[..] {
+            "aws_s3" => {
+                let s3_config = self.aws_s3.as_ref().expect("s3 config wasn't provided");
+                let client = create_client(&s3_config.region, &s3_config.auth, None, &cx.proxy)?;
+                let svc = self
+                    .build_s3_sink(&s3_config.options, client.clone(), cx)
+                    .map_err(|error| format!("{}", error))?;
+                Ok((svc, build_healthcheck(self.bucket.clone(), client)?))
+            }
+
+            service => Err(Box::new(ConfigError::UnsupportedService {
+                service: service.to_owned(),
+            })),
+        }
+    }
+
+    fn build_s3_sink(
+        &self,
+        s3_options: &S3Options,
+        client: S3Client,
+        cx: SinkContext,
+    ) -> std::result::Result<VectorSink, ConfigError> {
+        // we use lower default limits, because we send 100mb batches,
+        // thus no need in the the higher number of outcoming requests
+        let request_limits = self.request.unwrap_with(&DEFAULT_REQUEST_LIMITS);
+        let service = ServiceBuilder::new()
+            .settings(request_limits, S3RetryLogic)
+            .service(S3Service::new(client));
+
+        match s3_options.storage_class {
+            Some(class @ S3StorageClass::DeepArchive) | Some(class @ S3StorageClass::Glacier) => {
+                return Err(ConfigError::UnsupportedStorageClass {
+                    storage_class: format!("{:?}", class),
+                });
+            }
+            _ => (),
+        }
+
+        // We should avoid producing many small batches - this might slow down Log Rehydration,
+        // these values are similar with how DataDog's Log Archives work internally:
+        // batch size - 100mb
+        // batch timeout - 15min
+        let batch_size_bytes = NonZeroUsize::new(100_000_000);
+        let batch_size_events =
+            NonZeroUsize::new(1000).expect("batch size, in events, must be greater than 0"); //TODO what would be a reasonable default?
```

Comment:
This feels way too low if the goal is to hit multiple-tens-of-megabytes per batch.  Consider this real log line from my Plex server, which looks a lot like an Apache access log line:

    [23.12.132.121:55607] 200 GET /library/metadata/377031?asyncAugmentMetadata=1&checkFiles=1&includeChapters=1&includeConcerts=1&includeExternalMedia=1&includeExternalMetadata=1&includeExtras=1&includeGeolocation=0&includeMarkers=1&includeMeta=1&includeOnDeck=1&includePopularLeaves=1&includePreferences=0&includeRelated=0&includeRelatedCount=15&includeReviews=1

This line is long -- around 350 characters -- but 1000 of them would be less than 400KB.  You probably want something more like 100,000 events per batch, minimum.. maybe even 200,000.

Reply:
Yeah, it's definitely low, I just put a random number to think about later, thanks for spotting this.

Reply:
updated with 200_000(provided that logs on average are around 500bytes)